src/ngx_http_lua_balancer.c

```diff
@@ -322,105 +479,426 @@ ngx_http_lua_balancer_get_peer(ngx_peer_connection_t *pc, void *data)
         }
     }
 
-    if (bp->sockaddr && bp->socklen) {
+    if (ngx_http_lua_balancer_peer_set(bp)) {
         pc->sockaddr = bp->sockaddr;
         pc->socklen = bp->socklen;
+        pc->name = bp->host;
         pc->cached = 0;
         pc->connection = NULL;
-        pc->name = bp->host;
-
-        bp->rrp.peers->single = 0;
 
         if (bp->more_tries) {
             r->upstream->peer.tries += bp->more_tries;
         }
 
-        dd("tries: %d", (int) r->upstream->peer.tries);
+        if (ngx_http_lua_balancer_keepalive_enabled(bp)) {
+            ngx_http_lua_balancer_get_keepalive_pool(L, bp->kpool_crc32,
+                                                     &bp->kpool);
 
-        return NGX_OK;
-    }
+            if (bp->kpool == NULL
+                && ngx_http_lua_balancer_create_keepalive_pool(L, pc->log,
```

Comment:
How do you limit the number of pools created? I'm worried about too many connection pools getting created without bound. The variants of crc32 can be huge. How about doing a search with crc32 just with the existing keepalive upstream module?

Reply:
> How do you limit the number of pools created? 

In future PRs, I was planning on adding ways to retrieve metrics from the pools (e.g. number of active pools, total memory allocated to existing pools, shutting down existing pools, retrieve the number of connections by pool, `getreusedtimes()` per connection, etc...). It is then up to users to manage their upstream keepalive pools as desired, according to their own constraints.
Limiting the max number of pools via a simple number feels very limiting in comparison. We could introduce a new NGINX directive, but those are also extremely limiting since they greatly limit the potential for dynamicity in OpenResty-based applications. If anything, such a "limit" should be applicable dynamically imho (i.e. via a Lua API). But again, letting users interact with existing pools and decide for themselves how many pools they should maintain is much better imho.     

> How about doing a search with crc32 just with the existing keepalive upstream module?

I'm not sure I get what you mean here. We don't have access to the inner keepalive upstream module without an NGINX patch, which is what this PR aims at avoiding. Besides, the NGINX keepalive upstream module has many other limitations which resulted in this PR being created in the first place.

Reply:
It is worth noting that even our cosocket connection pooling does not implement such a limit on the total number of pools.

Reply:
I'm afraid it would still be tricky to do LRU by the user herself unless we also provide an API to return the least recently used pool name. Besides, doing a lot of metrics bookkeeping is extra overhead at runtime and extra complexity in the code. I also have mixed feeling about that.

Reply:
The OpenResty way of doing things is to impose default limits to avoid shooting oneself on the foot. We should at least have something.

Reply:
Granted, but how do today's users of the cosocket connection pooling API deal with this? They don't...

Reply:
@thibaultcha currently it's a single pool so there's no issue of pool leaking.

Reply:
For the first version, what's the current strategy (or reality) of pool lifetime management?

Reply:
Because it's common to add unbounded number of dynamic peers in the ngx_lua world. An unbounded number of pools created is a big concern.

Reply:
What I meant is that today, there is no `lua_socket_max_pools` directive imposing an upper bound on cosocket connection pools either, and, unless I am mistaken, we have not heard of prominent concerns in that regard. That said, I could add something to this PR if that would alleviate your concerns.